src/Jackalope/Version/VersionHandler.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+namespace Jackalope\Version;
+
+use Jackalope\Node;
+use Jackalope\Transport\TransportInterface;
+use PHPCR\UnsupportedRepositoryOperationException;
+
+class VersionHandler
+{
+    const MIX_VERSIONABLE = 'mix:versionable';
+    const MIX_SIMPLE_VERSIONABLE = 'mix:simpleVersionable';
+
+    /**
+     * @var TransportInterface
+     */
+    private $transport;
+
+    public function __construct(TransportInterface $transport)
+    {
+        $this->transport = $transport;
+    }
+
+    public static function addVersionProperties(Node $node, TransportInterface $transport)
+    {
+        // TODO solve without reflection
+        $reflection = new \ReflectionMethod(get_class($transport), 'storeNode');
+        $reflection->setAccessible(true);
+
+        // TODO move this to the version handler
+        $session = $node->getSession();
+        $rootNode = $session->getRootNode();
+
+        $node->setProperty('jcr:isCheckedOut', true);
+
+        if (!$rootNode->hasNode('jcr:system/jcr:versionStorage')) {
+            if (!$rootNode->hasNode('jcr:system')) {
+                $rootNode->addNode('jcr:system');
+            }
+            $rootNode->addNode('jcr:system/jcr:versionStorage');
+        }
+
+        $versionStorageNode = $rootNode->getNode('jcr:system/jcr:versionStorage');
+        if (!$versionStorageNode->hasNode($node->getIdentifier())) {
+            $versionStorageNode->addNode($node->getIdentifier(), 'nt:versionHistory');
+        }
+        $versionHistory = $versionStorageNode->getNode($node->getIdentifier());
+        $versionHistory->setProperty('jcr:versionableUuid', $node->getIdentifier());
+
+        // TODO Set jcr:copiedFrom if needed
+
+        $versionHistory->addNode('jcr:versionLabels', 'nt:versionLabels');
+        $rootVersion = $versionHistory->addNode('jcr:rootVersion', 'nt:version');
+        $reflection->invoke($transport, $rootVersion->getPath(), $rootVersion->getProperties());
```

Comment:
i would prefer to keep the architecture as it is now, but provide a handler that the client can delegate these things to, if the generic handler fits. that would keep things separated but allows to do backend specific performance tweaks.

the creation of infrastructure should also be in that handler rather than a VersionManager, or we will conflict with jackalope-jackrabbit which does not want to have any of this done on client side...